ambry-store/src/main/java/com.github.ambry.store/BlobStoreCompactor.java

```diff
@@ -0,0 +1,895 @@
+
+  /**
+   * Copies data from the provided log segment into the target log (swap spaces).
+   * @param logSegmentToCopy the {@link LogSegment} to copy from.
+   * @param duplicateSearchSpan the {@link FileSpan} in which to search for duplicates.
+   * @param startOffsetOfLastIndexSegmentForDelete  the start {@link Offset} of the index segment up until which delete
+   *                                                records are considered applicable.
+   * @return {@code true} if all the records in the log segment were copied. {@code false} if some records were not
+   * copied either because there was no more capacity or because a shutdown was initiated.
+   * @throws IOException if there were I/O errors during copying.
+   * @throws StoreException if there are any problems reading or writing to store components.
+   */
+  private boolean copyDataByLogSegment(LogSegment logSegmentToCopy, FileSpan duplicateSearchSpan,
+      Offset startOffsetOfLastIndexSegmentForDelete) throws IOException, StoreException {
+    for (Offset indexSegmentStartOffset : getIndexSegmentDetails(logSegmentToCopy.getName()).keySet()) {
+      IndexSegment indexSegmentToCopy = srcIndex.getIndexSegments().get(indexSegmentStartOffset);
+      if (needsCopying(indexSegmentToCopy.getEndOffset()) && !copyDataByIndexSegment(logSegmentToCopy,
+          indexSegmentToCopy, duplicateSearchSpan, startOffsetOfLastIndexSegmentForDelete)) {
+        // there is a shutdown in progress or there was no space to copy all entries.
+        return false;
+      }
+    }
+    return true;
+  }
+
+  /**
+   * Copies data in the provided {@code indexSegmentToCopy} into the target log (swap spaces).
+   * @param logSegmentToCopy the {@link LogSegment} to copy from.
+   * @param indexSegmentToCopy the {@link IndexSegment} that contains the entries that need to be copied.
+   * @param duplicateSearchSpan the {@link FileSpan} in which to search for duplicates.
+   * @param startOffsetOfLastIndexSegmentForDelete  the start {@link Offset} of the index segment up until which delete
+   *                                                records are considered applicable.
+   * @return {@code true} if all the records in the index segment were copied. {@code false} if some records were not
+   * copied either because there was no more capacity or because a shutdown was initiated.
+   * @throws IOException if there were I/O errors during copying.
+   * @throws StoreException if there are any problems reading or writing to store components.
+   */
+  private boolean copyDataByIndexSegment(LogSegment logSegmentToCopy, IndexSegment indexSegmentToCopy,
+      FileSpan duplicateSearchSpan, Offset startOffsetOfLastIndexSegmentForDelete) throws IOException, StoreException {
+    List<IndexEntry> allIndexEntries = new ArrayList<>();
+    // call into diskIOScheduler to make sure we can proceed (assuming it won't be 0).
+    diskIOScheduler.getSlice(INDEX_SEGMENT_READ_JOB_NAME, INDEX_SEGMENT_READ_JOB_NAME, 1);
```

Comment:
is there a TODO here? Didn't fully follow what `getSlice()` returns, but what does 1 signify? And the jobId does not seem to be used either?